extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -298,5 +352,74 @@ private void removeCookie(RoutingContext context, TenantConfigContext configCont
                 cookie.setPath(auth.cookiePath.get());
             }
         }
+        return cookie;
+    }
+
+    private boolean isLogout(RoutingContext context, TenantConfigContext configContext) {
+        Optional<String> logoutPath = configContext.oidcConfig.logout.path;
+
+        if (logoutPath.isPresent()) {
+            return context.request().absoluteURI().equals(
+                    buildUri(context, logoutPath.get()));
+        }
+
+        return false;
+    }
+
+    private SecurityIdentity trySilentRefresh(TenantConfigContext configContext, String idToken, String refreshToken,
+            RoutingContext context) {
+        OidcTenantConfig config = configContext.oidcConfig;
+
+        if (config.token.refreshTimeout.isPresent()) {
+            OAuth2AuthProviderImpl auth = ((OAuth2AuthProviderImpl) configContext.auth);
+            JWT jwt = auth.getJWT();
+            JsonObject tokenJson;
+
+            try {
+                tokenJson = jwt.decode(idToken);
+            } catch (Exception cause) {
+                throw new AuthenticationCompletionException(cause);
+            }
+
+            Long iat = tokenJson.getLong("iat");
+            long now = System.currentTimeMillis() / 1000;
+            Integer timeout = config.token.refreshTimeout.get();
+
+            if (now - iat >= timeout) {
+                try {
+                    if (!refreshingTokens.add(refreshToken)) {
```

Comment:
What is this `refreshingTokens` about, why would we want to have this cache ?

Reply:
Suppose you have the same user coming from different concurrent threads. This code tries to fail-fast for the last thread that tries to refresh the token if it is already happening.

Reply:
Hey, thanks, but IMHO it is not really preventive enough. It is recommended as a good practice to recycle the refresh tokens themselves as part of the RT grant (I'm def aware of such a practice :-) ), so what may happen is that a new RT is returned and we may end adding here that new token to this set while the other concurrent thread is already operating with the stale RT.
I don't know how to resolve it right but this cache can cause the problem IMHO.

Reply:
There is one thing though that I need to check. The refresh is a blocking operation, so I need to make sure it does not run in the IO Thread. Currently, we only run blocking if the tenant is not yet resolved, so we need to advertise somehow the `OidcIdentityProvider` or move the runblocking logic to the authentication mechanisms.

Reply:
I'm not following. I think you are talking about refresh token rotation ? So that you get a new one every time you refresh tokens ?

How would you add the new token if it was not yet sent to the browser?

Reply:
@pedroigor may be indeed do it there pre-emptively, in CodeAuthenticationMechansim

Reply:
I've also changed to use a bounded map so that we put a limit on how many refresh tokens we can handle simultaneously. Otherwise, we may have issues ...

I agree, maybe we can do the runblocking stuff there ...